### PR TITLE
BilbyCustomFunctions_Reduction.py script update

### PR DESCRIPTION
**Pull request by hortica**
BilbyCustomFunctions_Reduction.py script is a set of Bilby specific data reduction functions. It needs constant updates. To be able to use Mantid on several computers consistently and make users' life easier, the set of functions shall be properly updated, since it is included into the main repository. 
I believe current BilbyReductionScript.py shall be OK with this update, but the proper update of the latter to follow.

**Test BilbySANSDataProcessorTest.py developed earlier for Bilby shall be OK.**

*There is no associated issue, only functionality update.*

### CODE REVIEW

File: scripts/BilbyCustomFunctions_Reduction.py [hunk 1]
```diff
@@ -11,64 +11,61 @@
...
     if line == 'false':
-        bool_string = False
+       bool_string  = False
     elif line == 'true':
         bool_string = True
-    else:
-        print("Check value of {}".format(line))
-        print("It must be either True or False")
+    else:    
+        print "Check value of ",  line
```
Comment (TTitcombe) [inline scripts/BilbyCustomFunctions_Reduction.py]: for python 3 compatibility, we'll need to keep the print statements as they were and put `from __future__ import print_function` on line 7

**Comment by DTasev** [comment]
All of the `Flake8` errors seem to be from inconsistent/bad formatting. I suggest using an editor like [PyCharm (free edition)](https://www.jetbrains.com/pycharm/download/#section=linux) to fix that quickly. 

Alternatively you could [run the flake8 formatter manually](https://flake8.pycqa.org/en/latest/internal/formatters.html).

**Comment by hortica** [comment]
OK, thank you! Will fix till the end of the week.
Please advise. When the code will be updated, shall I create a new pull request?

**Comment by DTasev** [comment]
You don't have to, as soon as you make the changes and push it onto your `master` branch, it will be updated here automatically :slightly_smiling_face:

**Comment by hortica** [comment]
Cool. 
I am using VS Code, it seems Flake8 is available there. Apologise, forgot about beauty of a proper formatting, got too engaged with Git.
Thank you for your time!

**Comment by hortica** [comment]
> 
> 
> You don't have to, as soon as you make the changes and push it onto your `master` branch, it will be updated here automatically
Dimitar, question here.
How serious you are about 79-characters long lines? Shall I clean up long lines in my code?
Thank you!

**Comment by hortica** [comment]
> 
> 
> You don't have to, as soon as you make the changes and push it onto your `master` branch, it will be updated here automatically
Dimitar, question here.
How serious you are about 79-characters long lines? Shall I clean up long lines in my code?
Thank you!

**Comment by DTasev** [comment]
Our [flake8 configuration](https://github.com/mantidproject/mantid/blob/master/.flake8) allows lines up to 140 characters, otherwise the flake8 build will fail.

You can add custom flake8 args in VSCode with this setting: `"python.linting.flake8Args": ["--ignore=E303"]`, but it might be able to pick up our `.flake8` configuration file if the file is in the VSCode project

**Comment by hortica** [comment]
> Our [flake8 configuration](https://github.com/mantidproject/mantid/blob/master/.flake8) allows lines up to 140 characters, otherwise the flake8 build will fail.
> 
> You can add custom flake8 args in VSCode with this setting: `"python.linting.flake8Args": ["--ignore=E303"]`, but it might be able to pick up our `.flake8` configuration file if the file is in the VSCode project

Good day!
I think I've done with updates, I believe I made all changes,
https://github.com/hortica/mantid/commit/db37492bbd5229b02dc0102d623cbc901e995bc2

I am not sure if I can resolve a conflict I see.
Please advise what my next step is.
Thank you.
Anna.

**Comment by DTasev** [comment]
There's still some flake8 issues, you can see the errors [here](https://builds.mantidproject.org/job/pull_requests-flake8/19605/flake8/). If you have the [python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) it will prompt you to install `autopep8` when you format the file, if you don't have it. That should solve most of the flake8 issues.

For the merge conflict, it seems it's just an issue of the position of the function `get_pixel_size`. If you merge master of `mantidproject/mantid` in your branch, locally you will be able to see the merge conflict in the `scripts/BilbyCustomFunctions_Reduction.py` file. It seems you can also resolve it in Github, but I have not used that so I am not sure if it can commit to your branch.

**Comment by hortica** [comment]
Good day,
The conflict has been resolved on Saturday, then it seems the build failed because of some other issues, and then I've re-based everything just now.
//I use Flake8 plugin in Linter in VScode, and I've taken your settings. It seems it is working.
I believe I am good for now, till next failure :)

**Comment by hortica** [comment]
> 
> 
> have you had a chance to see why the build is failing?

Good day,
Is it a question to me?
I don't know, I see "All checks have passed".
Please advise what I can double check to reply.
Anna

**Comment by hortica** [comment]
> 
> 
> have you had a chance to see why the build is failing?

Good day,
Is it a question to me?
I don't know, I see "All checks have passed".
Please advise what I can double check to reply.
Anna

**Comment by DanNixon** [comment]
@hortica You can just ignore that message, it is automatically sent if the build is failing and the PR hasn't been touched in a while.

@TTitcombe Are you OK to keep reviewing this?

**Comment by TTitcombe** [comment]
@DanNixon didn't realise I had been assigned, but I can continue reviewing this

File: scripts/BilbyCustomFunctions_Reduction.py [hunk 2]
```diff
@@ -115,12 +120,93 @@ def strip_NaNs(output_workspace, base_output_name):
 
     return base_output_name
 
+##############################################################################
+# 26 March 2019
+
+
+def output_header(external_mode, used_wl_range, ws_sample, sample_thickness,
```
Comment (TTitcombe) [inline scripts/BilbyCustomFunctions_Reduction.py]: This function is quite long and might benefit from being split up in multiple functions or separated by more comments, however it's in with the style of the file so I don't think there is a need to change.

**Review by TTitcombe — DISMISSED** [review]
PR is mostly code cleanup. I have a small comment on a function introduced in this PR but I'm happy to Approve as it currently is

File: scripts/BilbyCustomFunctions_Reduction.py [hunk 3]
```diff
@@ -115,12 +120,93 @@ def strip_NaNs(output_workspace, base_output_name):
 
     return base_output_name
 
+##############################################################################
+# 26 March 2019
```
Comment (DanNixon) [inline scripts/BilbyCustomFunctions_Reduction.py]: Please remove dates in comments, Git can be used to find out when things were done.

File: scripts/BilbyCustomFunctions_Reduction.py [hunk 4]
```diff
@@ -90,22 +93,24 @@ def files_to_reduce(parameters, evaluate_files):
...
-        evaluate_files_l = evaluate_files_list(evaluate_files)  # call funciton for retrieve the IDs list
+        evaluate_files_l = evaluate_files_list(evaluate_files)  # call function for retrieve the IDs list
         for parameter in parameters:
             if int(parameter['index']) in evaluate_files_l:
                 files_to_reduce.append(parameter)
 
     return files_to_reduce
 
+##############################################################################
+# For output file formatting #################################################
```
Comment (DanNixon) [inline scripts/BilbyCustomFunctions_Reduction.py]: Please use Python docstrings for documenting functions. This keeps documentation consistent.

(the `###` lines between functions are bad practice too, however they are already used in this file so I have no great opinion about removing them)

File: scripts/BilbyCustomFunctions_Reduction.py [hunk 5]
```diff
@@ -115,12 +120,93 @@ def strip_NaNs(output_workspace, base_output_name):
...
+            resolution = 'Resolution used for calculation of dQ: ' + str(format((-100 * resolution_value), '.2f')) + '%'
+        else:
+            resolution = 'Resolution taken as wavelength binning;' + '\n' + 'the value is set to ' + \
+                str(format(resolution_value, '.2f')) + '%'  # on linear scale, hence the dQ calculation is meaningless'
+        header.append(resolution)
+    else:
+        resolution = "Nominal resolution: 10%"
+        header.append(resolution)
+
+    L1 = 'L1: ' + str(format(float(ws_sample.run().getProperty("L1").value), '.3f')) + ' m'
```
Comment (DanNixon) [inline scripts/BilbyCustomFunctions_Reduction.py]: Python variable names should not use upper case letters.

**Review by DanNixon — CHANGES_REQUESTED** [review]
A few more code style points.

File: scripts/BilbyCustomFunctions_Reduction.py (hunk 5 above)
Comment (hortica) [reply scripts/BilbyCustomFunctions_Reduction.py]: > 
> 
> Python variable names should not use upper case letters.

OK, thank you. I'll make all corrections and update the pull request.
Anna

**Comment by hortica** [comment]
I believe requested changes are done and code looks good now.
Thank you for checking.

**Review by DanNixon — APPROVED** [review]
The failing doc test in unrelated to the changes here.

**Comment by hortica** [comment]
> 
> 
> The failing doc test in unrelated to the changes here.

OK

**Comment by martyngigg** [comment]
Sphinx warning is fixed on master.

**Comment by hortica** [comment]
> 
> 
> Sphinx warning is fixed on master.

I cannot believe this little script is finally in!! Thank you very much for your help and advises :) looking forward to see it in the next NightlyBuild :)
Anna

**Pull request merged**